mobius-core/src/test/java/com/spotify/mobius/MobiusLoopTest.java

```diff
@@ -374,6 +395,99 @@ public void dispose() {
     observer.assertStates("Firstinit->1");
   }
 
+  @Test
+  public void eventSourceMayEmitEventsBeforeReturningFromDispose() throws Exception {
+    // Event sources should be allowed to emit events before they return
+    // from their dispose() method.
+
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(update, effectHandler)
+            .eventSource(new EmitDuringDisposeEventSource(new TestEvent("foo")));
+
+    builder.startFrom("foo").dispose();
+  }
+
+  @Test
+  public void effectHandlerMayEmitEventsBeforeReturningFromDispose() throws Exception {
+    // Effect handlers should be allowed to emit events before they return
+    // from their dispose() method.
+
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(update, new EmitDuringDisposeEffectHandler());
+
+    builder.startFrom("foo").dispose();
+  }
+
+  @Test
+  public void eventCausingSynchronousModelChangeDuringDisposeIsSafe() throws Exception {
+    // Given that an event is emitted during disposal of the loop and that the
+    // event is handled synchronously, any model change that happens in the update
+    // function should be safe.
+
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(update, effectHandler)
+            .eventSource(new EmitDuringDisposeEventSource(new TestEvent("bar")))
+            .eventRunner(() -> immediateRunner)
+            .effectRunner(() -> immediateRunner);
+
+    builder.startFrom("foo").dispose();
+  }
+
+  @Test
+  public void eventCausingSynchronousModelChangeDuringDisposeDoesNotEmitToObservers()
+      throws Exception {
+    // Given that an event is emitted during disposal of the loop and that the
+    // event is handled synchronously, any model change that happens in the update
+    // function should be ignored and not emitted to observers.
+
+    final MobiusLoop.Builder<String, TestEvent, TestEffect> builder =
+        Mobius.loop(update, effectHandler)
+            .eventSource(new EmitDuringDisposeEventSource(new TestEvent("bar")))
+            .eventRunner(() -> immediateRunner)
+            .effectRunner(() -> immediateRunner);
+
+    mobiusLoop = builder.startFrom("foo");
+    mobiusLoop.observe(observer);
+    mobiusLoop.dispose();
+
+    observer.assertStates("init", "foo");
```

Comment:
If this test fails, will the next state be "bar"? Also, what happens if I call `mobiusLoop.getMostRecentModel()`? Right now my guess is it'll return "bar"

Reply:
I think maybe the 'synchronous' bit is part of what's throwing me off; there should be no difference in behaviour between sync and async. Verifying the async behaviour is harder to do deterministically, but from a specification perspective, they should be the same.

Reply:
yes the `"init"` is just included by mistake from `setup` in the recording observer, it's not relevant for this test. the test was written first but when i saw that "init" was missing when running i just though "oh right it comes from `init`" and forgot that i hadn't registered it in this test case - completely my fault.

This entire suite of tests is specifically trying to see what happen when models, events, and effects are emitted from the update function, event sources, and effect handlers, during `dispose()` of the loop.

This test case is meant to specifically test something that could only happen synchronously, and make sure it doesn't crash, deadlock, or anything other bad like that.

During disposal,  `"bar"` in this case is emitted from the event source, and if the update function actually receives that, it will cause a model to be emitted. Since we are in the middle of disposal, no model should actually be emitted because in principle that could cause something outside of the loop to get a callback that could be used to send new events to the loop. There is no way to know outside the loop in that callback that the loop is in the middle of being disposed so it seemed reasonable to me to not have the callback happen in the first place.